perceval/client.py

```diff
@@ -0,0 +1,236 @@
+                 max_retries_on_status=MAX_RETRIES_ON_STATUS, status_forcelist=DEFAULT_STATUS_FORCE_LIST,
+                 method_whitelist=DEFAULT_METHOD_WHITELIST, raise_on_redirect=DEFAULT_RAISE_ON_REDIRECT,
+                 raise_on_status=DEFAULT_RAISE_ON_STATUS, respect_retry_after_header=DEFAULT_RESPECT_RETRY_AFTER_HEADER,
+                 default_sleep_time=DEFAULT_SLEEP_TIME, headers=None):
+        self.base_url = base_url
+
+        self.max_retries = max_retries
+        self.max_retries_on_connect = max_retries_on_connect
+        self.max_retries_on_read = max_retries_on_read
+        self.max_retries_on_redirect = max_retries_on_redirect
+        self.max_retries_on_status = max_retries_on_status
+        self.status_forcelist = status_forcelist
+        self.method_whitelist = method_whitelist
+        self.raise_on_redirect = raise_on_redirect
+        self.raise_on_status = raise_on_status
+        self.respect_retry_after_header = respect_retry_after_header
+        self.default_sleep_time = default_sleep_time
+
+        self._create_http_session(headers)
+
+    def __del__(self):
+        self._close_http_session()
+
+    def _create_http_session(self, headers=None):
+        """
+        Create a http session and initialize the retry object
+
+        :param headers: list of session headers
+        """
+        self.session = requests.Session()
+
+        if headers:
+            self.session.headers.update(headers)
+
+        retries = Retry(total=self.max_retries,
+                        connect=self.max_retries_on_connect,
+                        read=self.max_retries_on_read,
+                        redirect=self.max_retries_on_redirect,
+                        status=self.max_retries_on_status,
+                        method_whitelist=self.method_whitelist,
+                        status_forcelist=self.status_forcelist,
+                        backoff_factor=self.default_sleep_time,
+                        raise_on_redirect=self.raise_on_redirect,
+                        raise_on_status=self.raise_on_status,
+                        respect_retry_after_header=self.respect_retry_after_header)
+
+        self.session.mount('http://', HTTPAdapter(max_retries=retries))
+        self.session.mount('https://', HTTPAdapter(max_retries=retries))
+
+    def _close_http_session(self):
+        """Close the http session"""
+
+        self.session.close()
+
+    def fetch(self, url, payload=None, headers=None, method=GET, stream=False):
+        """
+        Fetch the data from a given URL
+
+        :param url: link to the resource
+        :param payload: payload of the request
+        :param headers: headers of the request
+        :param method: type of request call (GET or POST)
+        :param stream: defer downloading the response body until the response content is available
+        """
+        return self.__send_request(url, payload, headers, method, stream)
```

Comment:
As this method doesn't do anything else than calling `_send_request`, they should be merged into one.